### ISSUE TITLE
Implement `textDocument/prepareRename` request to player nicer with other rename providers

### ISSUE THREAD

**Issue by vldmr-bus** [dotnet/vscode-csharp#6643]
I noticed that f2(or rename symbol from menu) on C# variables is producing "This symbol can not be renamed" now. After playing with various settings trying to restore the functionality I finally downgraded from v2.9.20 to v2.8.23, and renaming works again.

**Comment by dibarbet** [comment]
Can you set the `dotnet.server.trace` configuration to `Trace` and include the contents of the C# output window?

**Comment by vldmr-bus** [comment]
Attached is log right after reload to updating to version v2.9.20. Renaming stopped working right away 
[c#log.zip](https://github.com/dotnet/vscode-csharp/files/13287291/c.log.zip)

**Comment by dibarbet** [comment]
Thanks - looking at the logs nothing obvious is going wrong on the server side.  The project is loading correctly.  I'm also not seeing a rename request in the server at all - which likely means VSCode is failing to even send us the request.  One more followup if you wouldn't mind

1.  First, set the VSCode log level to trace via this command:
    <img width="478" alt="image" src="https://github.com/dotnet/vscode-csharp/assets/5749229/6c871c66-c125-4963-9d16-3142473a103f">
2.  Run the repro with the failed rename
3.  Share the contents of the Extension Host output window and C# LSP Trace Logs output window


Make sure to turn VSCode logging back to info when done.

**Comment by dibarbet** [comment]
Oh, and a screenshot of the editor where the rename is happening and the failure would also be useful, if you can get that.

**Comment by vldmr-bus** [comment]
Attached screens and logs for two versions
[logs+screens.zip](https://github.com/dotnet/vscode-csharp/files/13288586/logs%2Bscreens.zip)

**Comment by dibarbet** [comment]
Thanks for the screenshots and extra logs.  The screenshots have some additional clues - in the failed version the colorization is also wrong.  The types aren't colored green as they are in the working screenshot.  

This generally indicates that the C# extension doesn't have any project information about the file (an be the case when the project is still loading for example).  C# devkit is responsible for loading the project in this case, so I'll need to get logs from that extension.

Instead of asking for individual files - would you mind enabling the VSCode trace logging again
![image](https://github.com/dotnet/vscode-csharp/assets/5749229/2f8c3d98-4de9-4e55-b397-aeb45c3d996e)

And zipping up the entire contents of the extensions log folder?  You can find it by running this command:
<img width="477" alt="image" src="https://github.com/dotnet/vscode-csharp/assets/5749229/e293f300-9a19-460f-bec7-3f7acc082707">

Thanks much for your help in getting to the bottom of this :)

**Comment by vldmr-bus** [comment]
I am not sure why types are not colorized on failed screen, probably did it too fast after reloading. I made another one after waiting for types to be colorized. It is in attached archive together with extensions log folder:
[exthost.zip](https://github.com/dotnet/vscode-csharp/files/13289650/exthost.zip)

**Comment by dibarbet** [comment]
So from the extension host logs there are 0 invocations of rename to the c# extension (`ms-dotnettools.csharp`).  There are however multiple invocations of rename to an extension with id `vldmr-bus.shalldn`, for example:
```
2023-11-07 16:15:11.317 [trace] [vldmr-bus.shalldn] INVOKE provider 'Z=>Z.resolveRenameLocation(C.URI.revive(J),se,oe)'
2023-11-07 16:15:11.318 [trace] [vldmr-bus.shalldn] provider DONE after 0ms
2023-11-07 16:15:15.660 [trace] [vldmr-bus.shalldn] INVOKE provider 'Z=>Z.resolveRenameLocation(C.URI.revive(J),se,oe)'
2023-11-07 16:15:15.661 [trace] [vldmr-bus.shalldn] provider DONE after 0ms
```

Since the C# extension is not getting invoked at all, I am guessing that the `vldmr-bus.shalldn` is getting called for rename instead of the C# extension.  Does it work if you disable that extension?

**Comment by vldmr-bus** [comment]
Disabling shalldn extension restored renaming functionality with version v2.9.20 of c# extension. With r-enabling shalldn extension the problem returned.

So why it works correctly with version v2.8.23 of c#, but interacts badly between shalldn and version v2.9.20? Is there something that needs to be corrected in shalldn extension?

**Comment by dibarbet** [comment]
I would guess that `vldmr-bus.shalldn` is registering its capabilities for c# files (or all files).  Generally this is controlled by the [document selector](https://code.visualstudio.com/api/references/document-selector) that you provide when you register your extension.  

I found this code on github (not 100% sure if its the same extension) - https://github.com/vldmr-bus/shalldn/blob/main/shalldn-ls/client/src/extension.ts#L99C6-L99C20 which seems suspicious.  It looks like its registering for any file, no matter the language.  It should probably instead be registering for a specific language or file extension.

I have no idea why it works in v2.8.23 - we haven't changed anything on the c# extension side in regards to rename or document selectors.  Perhaps some timing changes in initialization cause the C# extension to be registered for rename after `vldmr-bus.shalldn`, so it ends up calling `vldmr-bus.shalldn` first.

**Comment by vldmr-bus** [comment]
The reference to github source that you found is correct, that is the extension in question. Registration for any file type is intended, as shalldn extension provide ability to rename and lookup symbols that may be embedded in comments of any language. 

The order of initialization does matter as I found by selectively disabling and re-enabling extensions in different order, because enabling does not require reload, and thus allows guaranteed order of initialization. 

So I found that it is last one to be enabled that gets renaming priority. But what also seems to be the case that initialization order depends on the order of installation or re-enabling of extensions: installed/enabled last is initialized last, and thus gets renaming priority.

Long story short, after messing with enabling/disabling I ended up with state where c# at latest version took over renaming, but prevents shalldn symbol renaming in .cs files. That is good enough for me, because navigation through symbol is still working from .cs to .shalldn files, and from there renaming of shalldn symbols works normally.

I will thus close this ticket as resolved.

Thanks!

**Comment by dibarbet** [comment]
Thanks for the update, glad to hear your unblocked for now at least.  I'm going to re-open this issue because there might be something we can do on the C# extension side.

It looks like VSCode will call `prepareRename` on all servers that provide it and choose the one that does not reject the request. 
See https://github.com/microsoft/vscode/blob/ea4d99921cc790d49194e897021faee02a1847f7/src/vs/editor/contrib/rename/rename.ts#L58-L72

However - C# at this point doesn't implement the `prepareRename` request - so I believe it automatically ignores it.  I think we could make this scenario work most of the time if:
1.  C# implements prepareRename to say it can rename in C# code, but not in comments (we don't support that at the moment).
2.  `vldmr-bus.shalldn` implements prepareRename to say it can only rename in comments, and rejects the `prepareRename` request in C# code otherwise.

Unfortunately I don't think there is a way to always priortize the C# extension for rename, see https://github.com/microsoft/vscode/issues/115354

Updating this issue to track implementing the [prepareRename](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_prepareRename)request on the C# server side.
